pkg/providers/instancetype/offering/provider.go

```diff
@@ -218,10 +235,39 @@ func (p *DefaultProvider) cacheKeyFromInstanceType(it *cloudprovider.InstanceTyp
 		&hashstructure.HashOptions{SlicesAsSets: true},
 	)
 	return fmt.Sprintf(
-		"%s-%016x-%016x-%d",
+		"%s-%016x-%016x",
 		it.Name,
 		zonesHash,
 		capacityTypesHash,
-		p.unavailableOfferings.SeqNum,
 	)
 }
+
+func (p *DefaultProvider) updateOfferingAvailability() {
+	for k, v := range p.cache.Items() {
+		var updatedOfferings []*cloudprovider.Offering
+		// Extract instance type name from cache key
+		itName := strings.Split(k, "-")[0]
+		for _, offering := range v.Object.([]*cloudprovider.Offering) {
+			capacityType := offering.CapacityType()
+			// unavailableOfferings only affects on-demand & spot offerings
+			if capacityType == karpv1.CapacityTypeOnDemand || capacityType == karpv1.CapacityTypeSpot {
+				zone := offering.Zone()
+				isUnavailable := p.unavailableOfferings.IsUnavailable(ec2types.InstanceType(itName), zone, capacityType)
+				hasPrice := offering.Price > 0.0
+				// A new offering is created to ensure that the previous offering is not modified while still in use
+				updatedOfferings = append(updatedOfferings, &cloudprovider.Offering{
+					Requirements: offering.Requirements,
+					Price:        offering.Price,
+					Available:    !isUnavailable && hasPrice,
```

Comment:
This is missing the `itZones.Has(zone)` condition from the canonical definition in `createOfferings()` since `itZones` is not available here. There might be some edge case where a price is available for an invalid `(instanceType, zone, capacityType)`-tuple, which would make this incorrect. An alternate solution would be to simply not construct an `Offering` for any `zone` where a `(capacityType, instanceType)`-pair is not valid, though that would also result in them being dropped from the metrics (which could also be a considered positive side-effect?)

Reply:
If we dropped these instance types, this would induced drift. Increasing disruption rate for users is a pretty big draw back: https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/controllers/nodeclaim/disruption/drift.go#L114

Reply:
It might be better to include itZones (and itName) in the cache record then, since that would also solve the issue with extracting the instance type name from the cache key.